drivers/flash/nrf_qspi_nor.c

```diff
@@ -0,0 +1,688 @@
+#define QSPI_IS_BLOCK_ALIGNED(_ofs) (((_ofs) & (QSPI_BLOCK_SIZE - 1U)) == 0)
+
+/**
+ * @brief Main configuration structure
+ */
+static struct qspi_nor_data qspi_nor_memory_data = {
+	.sem = Z_SEM_INITIALIZER(qspi_nor_memory_data.sem, 1, 1),
+	.sync = Z_SEM_INITIALIZER(qspi_nor_memory_data.sync, 0, 1),
+};
+
+/**
+ * @brief Converts NRFX return codes to the zephyr ones
+ */
+static inline int qspi_get_zephyr_ret_code(int res)
+{
+	switch (res) {
+	case NRFX_SUCCESS:
+		return 0;
+	case NRFX_ERROR_INVALID_PARAM:
+	case NRFX_ERROR_INVALID_ADDR:
+		return -EINVAL;
+	case NRFX_ERROR_INVALID_STATE:
+		return -ECANCELED;
+	case NRFX_ERROR_BUSY:
+	case NRFX_ERROR_TIMEOUT:
+	default:
+		return -EBUSY;
+	}
+}
+
+static inline struct qspi_nor_data *get_dev_data(struct device *dev)
+{
+	return dev->driver_data;
+}
+
+static inline void qspi_lock(struct device *dev)
+{
+	struct qspi_nor_data *dev_data = get_dev_data(dev);
+
+	k_sem_take(&dev_data->sem, K_FOREVER);
+}
+
+static inline void qspi_unlock(struct device *dev)
+{
+	struct qspi_nor_data *dev_data = get_dev_data(dev);
+
+	k_sem_give(&dev_data->sem);
+}
+
+static inline void qspi_wait_for_completion(struct device *dev)
+{
+	struct qspi_nor_data *dev_data = get_dev_data(dev);
+
+	k_sem_take(&dev_data->sync, K_FOREVER);
+}
+
+static inline void qspi_complete(struct device *dev)
+{
+	struct qspi_nor_data *dev_data = get_dev_data(dev);
+
+	k_sem_give(&dev_data->sync);
+}
+
+/**
+ * @brief QSPI handler
+ *
+ * @param event Driver event type
+ * @param p_context Pointer to context. Use in interrupt handler.
+ * @retval None
+ */
+static void qspi_handler(nrfx_qspi_evt_t event, void *p_context)
+{
+	struct device *dev = p_context;
+
+	if (event == NRFX_QSPI_EVENT_DONE) {
+		qspi_complete(dev);
+		qspi_unlock(dev);
+	}
+}
+
+
+/* QSPI send custom command */
+static int qspi_send_cmd(struct device *dev, const struct qspi_cmd *cmd)
+{
+	/* Check input parameters */
+	if (!cmd) {
+		return -EINVAL;
+	}
+
+	qspi_lock(dev);
+
+	nrf_qspi_cinstr_conf_t cinstr_cfg = {
+		.opcode = cmd->op_code,
+		.io2_level = true,
+		.io3_level = true,
+		.wipwait = false,
+		.wren = true,
+	};
+	cinstr_cfg.length = sizeof(cmd->op_code);
+	if ((cmd->tx_buf != 0) && (cmd->rx_buf != 0)) {
+		cinstr_cfg.length += cmd->tx_buf->len + cmd->rx_buf->len;
+	} else if ((cmd->tx_buf != 0) && (cmd->rx_buf == 0)) {
+		cinstr_cfg.length += cmd->tx_buf->len;
+	} else if ((cmd->tx_buf == 0) && (cmd->rx_buf != 0)) {
+		cinstr_cfg.length += cmd->rx_buf->len;
+	}
+
+	int res = nrfx_qspi_cinstr_xfer(&cinstr_cfg, cmd->tx_buf->buf,
+					cmd->rx_buf->buf);
+
+	qspi_unlock(dev);
+	return qspi_get_zephyr_ret_code(res);
+}
+
+/* QSPI erase */
+static int qspi_erase(struct device *dev, u32_t addr, u32_t size)
+{
+	/* Check input parameters */
+	if (!size) {
+		return -EINVAL;
+	}
+
+	int res = 0;
+	const struct qspi_nor_config *params = dev->config->config_info;
+
+	while (size) {
+		if (size == params->size) {
```

Comment:
`addr` is not checked here, also for sizes greater then the size of the chip the loop will try to do partial erases.

Reply:
I have added check condition.